### PR TITLE
OCPBUGS-37543,ARO-9391: Avoid Shared Access Key usage for Azure Storage Account when using Managed Identity based auth

### PR DESCRIPTION
**Pull request by rajdeepc2792**
Recreated PR for https://github.com/openshift/cluster-image-registry-operator/pull/1078

Fixes:-
https://issues.redhat.com/browse/OCPBUGS-37543
&
https://issues.redhat.com/browse/ARO-9391

Changes:-
For Azure based storage account operations the Image Registry Operator is still using Azure Storage Account Access Keys generated using the Storage Accounts Client for manual mode credentials(i.e Managed Identity/Federated Identity).
This PR adds a diversion flow in operator when using federated tokens for all non Azure Stack Cloud to perform below operations:-

When creating storage account, disable the Shared Key Access.
When interacting with storage container use RBAC based authentication instead of using Shared Access Keys.
Testing:-

Unit Test Case:- No new test cases are added for the changes(Need to be performed)
Testing on 4.15 Managed Identity Cluster:-
Edited operator's deployment object to point the image to custom image-registry-operator image.
Operator and Registry pods got created and stayed in Running status successfully.
Operator and Registry pod was able to perform the operations on Storage Account and Container without fetching the Access Keys
Changing the AccountName in imageregistry.operator config created a new storage account with Shared Key Access disabled and container was also created successfully.

### CODE REVIEW

File: pkg/storage/azure/azure.go [hunk 1]
```diff
@@ -708,9 +773,83 @@ func (d *driver) assureStorageAccount(cfg *Azure, infra *configv1.Infrastructure
 		}
 	}
 
+	if !isAzureStackCloud(d.Config.CloudName) && cfg.FederatedTokenFile != "" {
+		azClient, err := d.newAzClient(cfg, environment, tagset)
+		if err != nil {
+			return "", false, err
+		}
+		err = azClient.DisableStorageAccountAccessKeyAccess(d.Context, cfg.ResourceGroup, accountName)
```
Comment (rajdeepc2792) [inline pkg/storage/azure/azure.go:781]: @flavianmissi , does this part makes sense to you? This is for disabling the Access Key Access if the mode of authentication is FederatedTokenFile/Managed Identity.

**Comment by rajdeepc2792** [comment]
Around the failing `ci/prow/e2e-azure-operator` pipeline, when it was getting tested against ocp/4.17 it was passing but after the `Building release initial from a snapshot of ocp/4.18`, the VNET getting created do not have tags `'kubernetes.io_cluster.<CLUSTER_NAME>: owned/shared'`, needs an investigation from pipeline infra side what has changed.

**Comment by dkeohane** [comment]
@adambkaplan can you support on this?

**Comment by flavianmissi** [comment]
/retitle OCPBUGS-37543,ARO-9391: Avoid Shared Access Key usage for Azure Storage Account when using Managed Identity based auth
The e2e-azure-operator failure seem unrelated (these tests are known to flake sometimes).

**Comment by flavianmissi** [comment]
/lgtm

About testing @xiuwang, we'll need to cover (in Azure) storage account and container creation and deletion in clusters configured with federated tokens (workload identity) and regular clusters. Same for Azure Stack Hub. If you can, doing the same in ARO would be great too. We also need to ensure the image registry can correctly read and write to the container created by the operator. The PR description gives a good background of the changes but please ask any questions that may come to mind. Also good to know that we're planning to backport this all the way back to 4.14.

@rajdeepc2792 please add any thoughts you may have about testing!

**Comment by xiuwang** [comment]
> About testing @xiuwang, we'll need to cover (in Azure) storage account and container creation and deletion in clusters configured with federated tokens (workload identity) and regular clusters. Same for Azure Stack Hub. If you can, doing the same in ARO would be great too. We also need to ensure the image registry can correctly read and write to the container created by the operator. The PR description gives a good background of the changes but please ask any questions that may come to mind. Also good to know that we're planning to backport this all the way back to 4.14.
> 
> @rajdeepc2792 please add any thoughts you may have about testing!

Thanks the suggestion @flavianmissi , I have checked azure federated tokens (workload identity) and regular clusters. I will check Azure Stack Hub and ARO(need pr merged ,then get a nightly build) too.

**Pull request merged**